File: src/online/musicbrainz_lookup.cpp
#include "musicbrainz_lookup.h"
#include <QNetworkAccessManager>
#include <QNetworkRequest>
#include <QNetworkReply>
#include <QEventLoop>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonArray>
#include <QUrlQuery>
#include <QDebug>
#include <QSslError>

namespace metatag {

MusicBrainzLookup::MusicBrainzLookup(QObject *parent)
    : QObject(parent)
{
}

void MusicBrainzLookup::lookupRelease(const QString& artist, const QString& album)
{
    if (artist.isEmpty() && album.isEmpty()) {
        emit lookupError("Artist and album cannot both be empty");
        return;
    }
    
    QNetworkAccessManager *nam = new QNetworkAccessManager(this);
    QUrl url("https://musicbrainz.org/ws/2/release/");
    QUrlQuery query;
    QStringList queryParts;
    if (!artist.isEmpty())
        queryParts.append(QString("artist:\"%1\"").arg(artist));
    if (!album.isEmpty())
        queryParts.append(QString("release:\"%1\"").arg(album));
    query.addQueryItem("query", queryParts.join(" AND "));
    query.addQueryItem("fmt", "json");
    url.setQuery(query);
    
    QNetworkRequest request(url);
    request.setRawHeader("User-Agent", "MetaTag/1.0 (https://github.com/metatag)");
    request.setRawHeader("Accept", "application/json");
    request.setTransferTimeout(10000); // 10 seconds
    
    QNetworkReply *reply = nam->get(request);
    connect(reply, &QNetworkReply::sslErrors, this, [reply](const QList<QSslError> &errors) {
        QList<QSslError> ignorableErrors;
        for (const QSslError &error : errors) {
            // Only ignore self-signed certificates for development/testing
            if (error.error() == QSslError::SelfSignedCertificate ||
                error.error() == QSslError::SelfSignedCertificateInChain) {
                ignorableErrors.append(error);
            }
        }
        if (ignorableErrors.size() == errors.size()) {
            qWarning() << "Ignoring self-signed certificate errors:" << ignorableErrors;
            reply->ignoreSslErrors(ignorableErrors);
        } else {
            qWarning() << "SSL errors occurred and will not be ignored:" << errors;
        }
    });
    connect(reply, &QNetworkReply::finished, this, [this, reply]() {
        if (reply->error() != QNetworkReply::NoError) {
            emit lookupError(reply->errorString());
            reply->deleteLater();
            return;
        }
        
        QByteArray data = reply->readAll();
        reply->deleteLater();
        
        QJsonDocument doc = QJsonDocument::fromJson(data);
        if (doc.isNull()) {
            emit lookupError("Failed to parse MusicBrainz JSON response");
            return;
        }
        
        QJsonObject root = doc.object();
        QJsonArray releasesArray = root["releases"].toArray();
        QVector<QString> releases;
        for (const QJsonValue& val : releasesArray) {
            QJsonObject releaseObj = val.toObject();
            QString id = releaseObj["id"].toString();
            if (!id.isEmpty())
                releases.append(id);
        }
        
        emit releasesFetched(releases);
    });
}

}

File: python/metatag/online/musicbrainz_lookup.py
"""MusicBrainz lookup for release information."""

import json
from typing import List

from PySide6.QtCore import QObject, Signal, QUrl, QUrlQuery
from PySide6.QtNetwork import (
    QNetworkAccessManager,
    QNetworkRequest,
    QNetworkReply,
    QSslError,
)


class MusicBrainzLookup(QObject):
    """Lookup MusicBrainz releases by artist/album."""

    releases_fetched = Signal(list)  # list of release IDs
    lookup_error = Signal(str)

    def __init__(self, parent: QObject = None):
        super().__init__(parent)
        self._nam = QNetworkAccessManager(self)
        self._base_url = "https://musicbrainz.org"

    def set_base_url(self, url: str) -> None:
        """Set the base URL for MusicBrainz API requests."""
        if not url:
            self._base_url = "https://musicbrainz.org"
            return
        # Normalize: strip trailing slashes
        self._base_url = url.strip().rstrip("/")

    def lookup_release(self, artist: str, album: str) -> None:
        """Lookup releases for given artist and album."""
        if not artist and not album:
            self.lookup_error.emit("Artist and album cannot both be empty")
            return

        url_str = f"{self._base_url}/ws/2/release/"
        url = QUrl(url_str)
        query = QUrlQuery()
        query_parts = []
        if artist:
            query_parts.append(f'artist:"{artist}"')
        if album:
            query_parts.append(f'release:"{album}"')
        query.addQueryItem("query", " AND ".join(query_parts))
        query.addQueryItem("fmt", "json")
        url.setQuery(query)

        request = QNetworkRequest(url)
        request.setRawHeader(b"User-Agent", b"MetaTag/1.0 (https://github.com/metatag)")
        request.setRawHeader(b"Accept", b"application/json")

        reply = self._nam.get(request)
        reply.sslErrors.connect(self._handle_ssl_errors)
        reply.finished.connect(lambda: self._handle_reply_finished(reply))

    def _handle_ssl_errors(self, errors: List[QSslError]) -> None:
        """Handle SSL errors, ignore self-signed certificates."""
        ignorable = []
        for error in errors:
            if error.error() in (
                QSslError.SslError.SelfSignedCertificate,
                QSslError.SslError.SelfSignedCertificateInChain,
            ):
                ignorable.append(error)
        if len(ignorable) == len(errors):
            # All errors are ignorable
            self.sender().ignoreSslErrors(ignorable)

    def _handle_reply_finished(self, reply: QNetworkReply) -> None:
        """Handle finished network reply."""
        try:
            if reply.error() != QNetworkReply.NetworkError.NoError:
                self.lookup_error.emit(reply.errorString())
                return

            data = reply.readAll().data()
            if not data:
                self.lookup_error.emit("Empty response from MusicBrainz")
                return

            result = json.loads(data)
            releases = result.get("releases", [])
            release_info = []
            for release in releases:
                # FIX: guard empty artist-credit / label-info lists before [0]
                artist_credits = release.get("artist-credit") or []
                label_info     = release.get("label-info") or []
                info = {
                    "id":      release.get("id"),
                    "title":   release.get("title", ""),
                    "artist":  artist_credits[0].get("name", "") if artist_credits else "",
                    "year":    release.get("date", "")[:4],
                    "label":   label_info[0].get("label", {}).get("name", "") if label_info else "",
                    "country": release.get("country", ""),
                }
                release_info.append(info)

            if not release_info:
                self.lookup_error.emit("No releases found for artist/album")
                return

            self.releases_fetched.emit(release_info)

        except json.JSONDecodeError as e:
            self.lookup_error.emit(f"Failed to parse MusicBrainz JSON: {e}")
        except Exception as e:
            self.lookup_error.emit(f"Unexpected error: {e}")

    def fetch_release_details(self, release_id: str) -> None:
        """Fetch detailed information (tracklist) for a specific release."""
        url_str = f"{self._base_url}/ws/2/release/{release_id}"
        url = QUrl(url_str)
        query = QUrlQuery()
        query.addQueryItem("inc", "recordings+artist-credits")
        query.addQueryItem("fmt", "json")
        url.setQuery(query)

        request = QNetworkRequest(url)
        request.setRawHeader(b"User-Agent", b"MetaTag/1.0 (https://github.com/metatag)")
        reply = self._nam.get(request)
        reply.finished.connect(lambda: self._handle_details_reply(reply))

    def _handle_details_reply(self, reply: QNetworkReply) -> None:
        """Parse the detailed release JSON."""
        try:
            if reply.error() != QNetworkReply.NetworkError.NoError:
                self.lookup_error.emit(reply.errorString())
                return
            
            data = json.loads(reply.readAll().data())
            tracks = []
            for medium in data.get("media", []):
                for track in medium.get("tracks", []):
                    # FIX: guard empty artist-credit before [0] access
                    t_credits = track.get("artist-credit") or []
                    tracks.append({
                        "title":  track.get("title", ""),
                        "artist": t_credits[0].get("name", "") if t_credits else "",
                        "number": track.get("position", ""),
                    })

            # FIX: guard release-level artist-credit before [0] access
            r_credits = data.get("artist-credit") or []
            self.releases_fetched.emit([{
                "title":  data.get("title"),
                "artist": r_credits[0].get("name", "") if r_credits else "",
                "tracks": tracks,
            }])
        except Exception as e:
            self.lookup_error.emit(f"Details error: {e}")
        finally:
            reply.deleteLater()
